src/io/network/addrinfo.hpp

```diff
@@ -11,21 +11,39 @@
 
 #pragma once
 
+#include <netdb.h>
+#include <string>
+
 namespace io::network {
 
+struct Endpoint;
+
 /**
  * Wrapper class for getaddrinfo.
  * see: man 3 getaddrinfo
  */
 class AddrInfo {
-  explicit AddrInfo(struct addrinfo *info);
-
  public:
+  struct Iterator {
```

Comment:
Hm, yes, I'm aware - I actually wanted to keep the code as small as possible by only implementing the bare minimum to make range based for working, but I can add the full forward iterator interface.

Reply:
Thinking of it, only an input iterator is required.

Reply:
Two things here. I could be fine for "just making it work" for something more complex (like using runtime polymorphism vs static -- because the  second requires CRTP and a little bit more templated work for a small gain in performance for most of the time). However, for this case we opted in to make AddrInfo iterable which means that we either do that and implement the whole interface (since we opted in for that) or not at all. Half baked glue code should be avoided. 

For the input iterator, I am not sure, I quote from cppreference on LegacyInputIterator:
>once a LegacyInputIterator i has been incremented, all copies of its previous value may be invalidated.
Is that the case for `AddrInfo` ? If the answer is yes, then `InputIterator` suffices for this  case.